Platform/OpenNtfsDxe/Open.c

```diff
@@ -0,0 +1,347 @@
+/** @file
+  Copyright (c) 2022, Mikhail Krichanov. All rights reserved.
+  SPDX-License-Identifier: BSD-3-Clause
+
+  Functional and structural descriptions follow NTFS Documentation
+  by Richard Russon and Yuval Fledel
+**/
+
+#include "NTFS.h"
+#include "Helper.h"
+
+EFI_STATUS
+EFIAPI
+FileOpen (
+  IN EFI_FILE_PROTOCOL  *This,
+  OUT EFI_FILE_PROTOCOL **NewHandle,
+  IN CHAR16             *FileName,
+  IN UINT64             OpenMode,
+  IN UINT64             Attributes
+  )
+{
+  EFI_STATUS     Status;
+  EFI_NTFS_FILE  *File;
+  EFI_FS         *FileSystem;
+  EFI_NTFS_FILE  *NewFile;
+  CHAR16         Path[MAX_PATH];
+  CHAR16         CleanPath[MAX_PATH];
+  CHAR16         *DirName;
+  INTN           Index;
+  INTN           Length;
+  BOOLEAN        AbsolutePath;
+
+  ASSERT (This != NULL);
+  ASSERT (NewHandle != NULL);
+  ASSERT (FileName != NULL);
+
+  File = (EFI_NTFS_FILE *) This;
+  FileSystem = File->FileSystem;
+  AbsolutePath = (*FileName == L'\\');
+
+  if (OpenMode != EFI_FILE_MODE_READ) {
+    DEBUG ((DEBUG_INFO, "NTFS: File '%s' can only be opened in read-only mode\n", FileName));
+    return EFI_WRITE_PROTECTED;
+  }
+
+  if ((*FileName == 0) || (CompareMem (FileName, L".", 2) == 0)) {
+    File->RefCount++;
+    *NewHandle = This;
+    return EFI_SUCCESS;
+  }
+
+  if (AbsolutePath) {
+    Length = 0;
+  } else {
+    Status = StrCpyS (Path, MAX_PATH, File->Path);
+    if (EFI_ERROR (Status)) {
+      DEBUG ((DEBUG_INFO, "NTFS: Could not copy string.\n"));
+      return Status;
+    }
+
+    Length = StrnLenS (Path, MAX_PATH);
+    if ((Length == 0) || (Path[Length-1] != '/')) {
+      Path[Length++] = L'/';
```

Comment:
Assuming ``File->Path`` is L"", then ``Length == 0``, and ``Path = { L'\0', <undef> }``. We enter, and set ``Path[0] = L' /'``, such that we get ``Path = { L'/', <undef> }``. Why is it guaranteed to be terminated here?

Reply:
Sorry, it's not guaranteed, but it's taken care of later. Comments when unsafe operations are intentional would really help